core/src/main/scala/org/bykn/bosatsu/Normalization.scala

```diff
@@ -0,0 +1,666 @@
+                    (fh(vh, env._1), env._2) match {
+                      case (_, NoMatch) => NoMatch
+                      case (NoMatch, _) => NoMatch
+                      case (Matches(env1), Matches(_)) => loop(vt, ft, (env1, Matches(env1)))
+                      case (Matches(env1), NotProvable) => loop(vt, ft, (env1, NotProvable))
+                      case (NotProvable, _) => loop(vt, ft, (env._1, NotProvable))
+                    }
+                  case Nil => env._2 // mismatch in size, shouldn't happen statically
+                }
+            }
+          loop(as, itemFns, (acc, Matches(acc)))
+        }
+
+        maybeIdx match {
+          case None => 
+            // this is a struct, which means we expect it
+            { (arg: NormalExpression, acc: PatternEnv) =>
+              arg match {
+                case NormalExpression.Struct(_, args) =>
+                  processArgs(args, acc)
+                case _ =>
+                  NotProvable
+              }
+            } 
+
+          case Some(idx) =>
+            // we don't check if idx < 0, because if we compiled, it can't be
+            val result = { (arg: NormalExpression, acc: PatternEnv) =>
+              arg match {
+                case NormalExpression.Struct(enumId, args) =>
+                  if (enumId == idx) processArgs(args, acc)
+                  else NoMatch
+                case _ =>
+                  NotProvable
+              }
+            }
+            result
+        }
+  }
+
+  def findMatch(m: NormalExpression.Match) =
+    m.branches.collectFirst(Function.unlift( { case (pat, result) =>
+      maybeBind(pat).apply(m.arg, Map()) match {
+        case Matches(env) => Some(Some((pat, env, result)))
+        case NotProvable => Some(None)
+        case NoMatch => None
+      }
+    })).get // Totallity of matches should ensure this will always find something unless something has gone terribly wrong
+
+  def solveMatch(env: PatternEnv, result: NormalExpression) =
+    ((env.size - 1) to 0 by -1).toList.map(env.get(_).get) // If this exceptions then somehow we didn't get enough names in the env
+      .foldLeft(result) { case (ne, arg) => NormalExpression.App(ne, arg) }
+ 
+  def normalOrderReduction(expr: NormalExpression): NormalExpression = {
+    import NormalExpression._
+    val nextExpr = expr match {
+      case App(Lambda(nextExpr), arg) => {
+        applyLambdaSubstituion(nextExpr, Some(arg), 0)
+      }
+      case m@Match(struct@Struct(enum, args), branches) =>
+        findMatch(m) match {
+          case None => m
+          case Some((pat, env, result)) => solveMatch(env, result)
+        }
+      case Recursion(Lambda(innerExpr)) if(innerExpr.maxLambdaVar.map(_ < 0).getOrElse(true)) => {
+        applyLambdaSubstituion(innerExpr, None, 0)
+      }
+      case Lambda(App(innerExpr, LambdaVar(0))) => {
+        innerExpr
+      }
+      case _ => expr
+    }
+    val res = nextExpr match {
```

Comment:
why the two steps? why can't we express in a single loop? Can you comment?

Reply:
I keep convincing myself that I need this but then when I go to write out the explanation I convince myself of the opposite. I should probably construct a proof one way or the other.

The intuition is that I'm looking for the opportunity to reduction the head without reducing the children first (which would be applicative order reduction I think). So first you look for an opportunity to reduce the head. But the new head could be anything, so you have to match through all expression types again.

This gets you to the second step, which is to reduce the children only if their are no opportunities on the head. The second step consists of,
  1) checking for a head reduction opportunity. Recurs if it exists
  2) fully reduce the children
  3) check if the children reduction created a new head reduction opportunity. Recurs if it exists

(I should add the above as a comment)

All this said, a lot of this code feels like exactly what dagon does